File: src/mission_control_lib/mission_control_lib/mission_control.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import Int8
from rcl_interfaces.msg import SetParametersResult
from mission_control_lib.subscription_manager import MISSION_ANNOUNCE_TOPIC_NAME, MISSION_ANNOUNCE_QOS

"""
Authored By Chris (T0fu) Lobo

Missions indicated in a mission_array, with a 1 representing the topic will be subscribed to, while 0 means the topic is destroyed for that mission.
The missions are indexed as follows:
    0: "TESTING",
    1: "LAP 2 ONWARDS",
    2: "INSPECTION",
    3: "ACTUATOR CHECK",
    4: "MANUAL CONTROL",
    5: "AUTONOMOUS STATIC",
    6: "ACCELERATION",
    7: "SKIDPAD",
    8: "TRACKDRIVE",
    9: "EMERGENCY STOP",
    10: "MISSION COMPLETE"

The nodes will by default start in the TESTING mission. So ensure if you have any nodes using this library, that it be set to 1 in index 0
if you wish to test.

"""

class MissionControl(Node):
    def __init__(self):
        super().__init__('MissionControl')

        self.declare_parameter('current_mission', 0)
        self.current_mission = self.get_parameter('current_mission').get_parameter_value().integer_value
        self.add_on_set_parameters_callback(self.on_param_change)

        self.mission_publisher = self.create_publisher(Int8, MISSION_ANNOUNCE_TOPIC_NAME, MISSION_ANNOUNCE_QOS)

    def on_param_change(self, params):
        for param in params:
            if param.name == 'current_mission':
                self.current_mission = param.value
                self.publish_signals()
                self.get_logger().info(f"Mission changed to {param.value}")
        return SetParametersResult(successful=True)

    def publish_signals(self):
        msg = Int8()
        msg.data = self.current_mission
        self.mission_publisher.publish(msg)

def main(args=None):
    rclpy.init(args=args)
    node = MissionControl()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/mission_control_cpp/src/mission_control.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/int8.hpp>
#include <rcl_interfaces/msg/set_parameters_result.hpp>
#include "mission_control_cpp/subscription_manager.hpp"

class MissionControl : public rclcpp::Node
{
public:
    MissionControl() : Node("MissionControl")
    {
        this->declare_parameter<int>("current_mission", 0);
        current_mission_ = this->get_parameter("current_mission").as_int();

        mission_publisher_ = this->create_publisher<std_msgs::msg::Int8>(
            MISSION_ANNOUNCE_TOPIC_NAME, MISSION_ANNOUNCE_QOS);

        parameter_callback_handle_ = this->add_on_set_parameters_callback(
            std::bind(&MissionControl::on_param_change, this, std::placeholders::_1));
    }

private:
    int current_mission_;
    rclcpp::Publisher<std_msgs::msg::Int8>::SharedPtr mission_publisher_;
    rclcpp::node_interfaces::OnSetParametersCallbackHandle::SharedPtr parameter_callback_handle_;

    rcl_interfaces::msg::SetParametersResult on_param_change(
        const std::vector<rclcpp::Parameter>& params)
    {
        for (const auto& param : params)
        {
            if (param.get_name() == "current_mission")
            {
                current_mission_ = param.as_int();
                publish_signals();
                RCLCPP_INFO(this->get_logger(), "Mission changed to %d", current_mission_);
            }
        }
        rcl_interfaces::msg::SetParametersResult result;
        result.successful = true;
        return result;
    }

    void publish_signals()
    {
        std_msgs::msg::Int8 msg;
        msg.data = current_mission_;
        mission_publisher_->publish(msg);
    }
};

int main(int argc, char** argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<MissionControl>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
